torchprime/torch_xla_models/train.py

```diff
@@ -305,14 +360,13 @@ def main(config: DictConfig):
   tokenizer_name = config.model.tokenizer_name
   tokenizer = AutoTokenizer.from_pretrained(tokenizer_name)
 
-  model = initialize_model_class(config.model)
+  # Set the model dtype to bfloat16
+  with set_default_dtype(torch.bfloat16), torch_xla.device():
+    model = initialize_model_class(config.model)
+
   n_params = sum([p.numel() for p in model.parameters()])
   logger.info(f"Training new model from scratch - Total size={n_params} params")
 
-  # Set the model dtype to bfloat16
-  model = model.to(torch.bfloat16)
-  model = model.to("xla")
```

Comment:
nit: Is the model initialized directly on the device?

Reply:
Yes, now the model is initialized directly on the TPU. For smaller models this shouldn't affect things. For large models like Llama 3.1 405B, it's actually impossible to initialize the model on the CPU (due to insufficient host memory) so the TPU route is the only option, short of hacks like initializing every layer manually.

Reply:
For Llama 405B on 2 pods, I had to add the `--xla_tpu_spmd_rng_bit_generator_unsafe=true` XLA flag to enable sharding propagation on the RNG. For smaller models, we could probably do without this flag.

Reply:
I am a little confused regarding how that happens. initialize_model_class initialize the model and none of the weight initializations here specify the device. It seems like the sharding code moves the tensors to the device. Can you provide some clarity here?

https://github.com/AI-Hypercomputer/torchprime/blob/main/torchprime/torch_xla_models/llama/model.py#L200

Reply:
Yes. The `with set_default_dtype(torch.bfloat16), torch_xla.device():` context manager does two things:
- Set the default dtype to bf16
- Set the default device to the XLA device

Under this context manager, all the weights will be initialized on the XLA device. When we run `nn.Linear(...)` under the `with torch_xla.device():` context manager, the `nn.Linear(...)` constructor gets captured into LazyTensor graph. No actual initialization takes place. This way, we can `xs.mark_sharding` individual weights to specify _post-hoc_ that those weights get sharded/placed in a specific way. Finally, a `torch_xla.sync()` call executes this whole graph. In summary, this is a very clean way to initialize the model weights in a sharded way by capturing the model constructor into HLO.

Reply:
Ah, makes sense!